indico/modules/events/registration/controllers/display.py

```diff
@@ -308,6 +308,24 @@ def success_url(self):
         return url_for('.display_regform', self.registration.locator.registrant)
 
 
+class RHRegistrationWithdraw(RHRegistrationFormRegistrationBase):
+    """Withdraw a registration"""
+
+    def _process(self):
+        if not self.registration.can_be_modified:
+            if not self.registration.registration_form.is_modification_open:
+                flash(_("The modification/withdrawal period is over"), "warning")
+            elif self.registration.registration_form.modification_mode.name == 'allowed_until_payment':
+                flash(_("Withdrawal is not allowed after payment"), "warning")
+            else:
+                flash(_("The registration cannot be modified/withdrawn"), "warning")
+            return redirect(url_for('.display_regform', self.registration.locator.registrant))
+
+        self.registration.state = RegistrationState.withdrawn
```

Comment:
This should use [`update_state()`](https://github.com/indico/indico/blob/9955b2ab920ac8af753937e79a038b78754ba4d7/indico/modules/events/registration/models/registrations.py#L458) (with a new kwarg for withdrawal) so the `registration_state_updated` signal gets triggered.